In [21]:
!pip install pandas pyxlsb

In [1]:
import pandas as pd
import numpy as np

In [2]:
file_path = "/Users/karenaraque/Desktop/practica_cartera_temu/DataFramePrueba 2025_08.xlsb"
# Lista de hojas para ver cuáles tiene
excel_hojas = pd.ExcelFile(file_path, engine="pyxlsb")
print(excel_hojas.sheet_names)

['DataFramePrueba', 'diccionario']


In [3]:
clientes = pd.read_excel(file_path, sheet_name="DataFramePrueba", engine="pyxlsb")
clientes.head(10)

IdentificadorCliente               FechaEvento         UsabilidadCupo  \
0                     1  2022-09-19T13:25:31.867Z  0.1184320077740548479   
1                     2  2023-08-23T11:33:46.417Z  0.0771590000000000000   
2                     3  2022-10-01T14:59:48.920Z  0.1855666666666666667   
3                     4  2022-09-22T21:25:09.187Z                   null   
4                     5  2023-03-19T17:48:52.310Z  0.4752148710773535878   
5                     6  2022-11-10T22:14:06.450Z  0.1561200000000000000   
6                     7  2022-12-04T21:33:38.130Z  0.5927940000000000000   
7                     8  2022-06-23T09:44:55.610Z  0.8576124728323314114   
8                     9  2022-10-01T13:32:38.053Z  0.2435427485988745499   
9                    10  2023-09-26T13:59:00.617Z  0.9671000000000000000   

    CategoriaPrincipalCredito  DiasMaximosMoraCreditosGenerados  \
0                         NaN                               0.0   
1             hogar-y-muebles                               NaN   
2                         NaN                             129.0   
3                         NaN                               0.0   
4                 computacion                               0.0   
5                         NaN                              49.0   
6  belleza-y-cuidado-personal                               0.0   
7                         NaN                               0.0   
8                         NaN                              31.0   
9             hogar-y-muebles                               NaN   

   NumeroCreditosGPrevius  NumeroCreditosGCanalFPrevius  \
0                     9.0                           9.0   
1                     5.0                           4.0   
2                    33.0                          33.0   
3                     3.0                           3.0   
4                     4.0                           4.0   
5                    18.0                          18.0   
6                     2.0                           2.0   
7                    12.0                          11.0   
8                    14.0                          14.0   
9                     2.0                           2.0   

   NumeroCreditosGEstadoActivosPrevius  NumeroCreditosGEstadoPagadosPrevius  \
0                                  0.0                                  9.0   
1                                  1.0                                  4.0   
2                                  0.0                                 33.0   
3                                  0.0                                  3.0   
4                                  0.0                                  4.0   
5                                  0.0                                 18.0   
6                                  0.0                                  2.0   
7                                  0.0                                 12.0   
8                                  0.0                                 14.0   
9                                  0.0                                  2.0   

   NumeroCreditosGCanalVPrevius  ...  TipoMunicipioEntregaTC  \
0                           0.0  ...               PRINCIPAL   
1                           1.0  ...              INTERMEDIO   
2                           0.0  ...                PEQUEÑO   
3                           0.0  ...                  GRANDE   
4                           0.0  ...                  GRANDE   
5                           0.0  ...               PRINCIPAL   
6                           0.0  ...                 VIRTUAL   
7                           1.0  ...                 POBLADO   
8                           0.0  ...              INTERMEDIO   
9                           0.0  ...              INTERMEDIO   

   CanalMunicipioEntregaTC  NumeroIntentosFallidos  CupoAprobado  UsoAppWeb  \
0                   Fisico                     0.0  3.000000e+10        NaN   
1                   Fisico                   

In [4]:
diccionario= pd.read_excel(file_path, sheet_name="diccionario", engine="pyxlsb")
diccionario.head(30)


IdentificadorCliente  \
0                              FechaEvento   
1                           UsabilidadCupo   
2                CategoriaPrincipalCredito   
3         DiasMaximosMoraCreditosGenerados   
4                   NumeroCreditosGPrevius   
5             NumeroCreditosGCanalFPrevius   
6             NumeroCreditosGCanalVPrevius   
7      NumeroCreditosGEstadoActivosPrevius   
8      NumeroCreditosGEstadoPagadosPrevius   
9                   NumeroCreditosLPrevius   
10     NumeroCreditosLEstadoActivosPrevius   
11     NumeroCreditosLEstadoPagadosPrevius   
12                 FechaVinculacionCliente   
13                          FechaPrimerUso   
14  TotalPagosEfectuadosGlobalmentePrevius   
15   TotalPagosEfectuadosLocalmentePrevius   
16                  CodigoAlmacenEntregaTC   
17                CodigoMunicipioEntregaTC   
18                  TipoMunicipioEntregaTC   
19                 CanalMunicipioEntregaTC   
20                  NumeroIntentosFallidos   
21                            CupoAprobado   
22                               UsoAppWeb   
23                         ScoreCrediticio   
24                                  Genero   
25                                    Edad   
26                          PerdidaCartera   
27                                DiasMora   

           Numero unico que identifica a cada cliente  
0   Fecha en la cual se hizo la compra en cuestión...  
1   Division entre el valor de la compra y el cupo...  
2   Categoría mayoritaria de la compra que quiere ...  
3   En la historia que lleva el cliente sacando cr...  
4   Cantidad de creditos que ha hecho el cliente a...  
5   Cantidad de creditos que ha hecho el cliente a...  
6   Cantidad de creditos que ha hecho el cliente a...  
7   Cantidad de creditos que ha hecho el cliente a...  
8   Cantidad de creditos que ha hecho el cliente a...  
9   Cantidad de creditos que ha hecho el cliente a...  
10  Cantidad de creditos que ha hecho el cliente e...  
11  Cantidad de creditos que ha hecho el cliente e...  
12  Fecha en la cual se le otorgó la tarjeta de cr...  
13  Fecha en la cual el cliente usó por primera ve...  
14  Pagos que ha realizado el cliente a la tarjeta...  
15  Pagos que ha realizado el cliente a la tarjeta...  
16  Almacen en el cual se le otorgó la tarjeta de ...  
17  municipio en el cual se le otorgó la tarjeta d...  
18  Tipo municipio en el cual se le otorgó la tarj...  
19  Canal por el cual se le otorgó la tarjeta de c...  
20  Cuantas veces el cliente pidió la Tarjeta de C...  
21                      Cupo de la tarjeta de credito  
22  Cual de los canales digitales usa mas el clien...  
23  Puntaje general que califica a un cliente en t...  
24                                 Genero del cliente  
25                                   Edad del cliente  
26  Es el valor que se quiere predecir. 1:El clien...  
27  Indica los dias de Mora que tiene el cliente, ...

# Tipos de datos por columna

In [5]:
print("Shape:", clientes.shape)
print("\nTipos de datos:")
print(clientes.dtypes)

Shape: (146939, 30)

Tipos de datos:
IdentificadorCliente                        int64
FechaEvento                                object
UsabilidadCupo                             object
CategoriaPrincipalCredito                  object
DiasMaximosMoraCreditosGenerados          float64
NumeroCreditosGPrevius                    float64
NumeroCreditosGCanalFPrevius              float64
NumeroCreditosGEstadoActivosPrevius       float64
NumeroCreditosGEstadoPagadosPrevius       float64
NumeroCreditosGCanalVPrevius              float64
NumeroCreditosLPrevius                    float64
NumeroCreditosLEstadoActivosPrevius       float64
NumeroCreditosLEstadoPagadosPrevius       float64
FechaVinculacionCliente                   float64
FechaPrimerUso                            float64
FechaUltimoUso                            float64
TotalPagosEfectuadosGlobalmentePrevius    float64
TotalPagosEfectuadosLocalmentePrevius     float64
CodigoAlmacenEntregaTC                     object
CodigoMunicip

# Valores únicos (para ver si son numéricas o categóricas, si hay códigos raros)

In [6]:
# Muestra hasta 15 valores únicos por columna
for col in clientes.columns:
    uniques = clientes[col].unique()
    nuniques = len(uniques)
    print(f"\n{col} — {nuniques} valores únicos")
    print(uniques[:15])  # muestra los primeros 15 valores distintos



IdentificadorCliente — 146939 valores únicos
[ 1  2  3  4  5  6  7  8  9 10 11 12 13 14 15]

FechaEvento — 146939 valores únicos
['2022-09-19T13:25:31.867Z' '2023-08-23T11:33:46.417Z'
 '2022-10-01T14:59:48.920Z' '2022-09-22T21:25:09.187Z'
 '2023-03-19T17:48:52.310Z' '2022-11-10T22:14:06.450Z'
 '2022-12-04T21:33:38.130Z' '2022-06-23T09:44:55.610Z'
 '2022-10-01T13:32:38.053Z' '2023-09-26T13:59:00.617Z'
 '2023-07-12T12:09:24.737Z' '2022-05-31T09:22:36.477Z'
 '2022-05-11T15:22:39.510Z' '2023-08-31T18:23:56.993Z'
 '2022-10-01T22:10:17.433Z']

UsabilidadCupo — 107705 valores únicos
['0.1184320077740548479' '0.0771590000000000000' '0.1855666666666666667'
 'null' '0.4752148710773535878' '0.1561200000000000000'
 '0.5927940000000000000' '0.8576124728323314114' '0.2435427485988745499'
 '0.9671000000000000000' '0.1248160000000000000' '0.3081000000000000000'
 '0.1855000000000000000' '0.0732485183968226177' '0.1845000000000000000']

CategoriaPrincipalCredito — 32 valores únicos
[nan 'hogar-y-mueble

# % de nulos

In [7]:
print("Duplicados:", clientes.duplicated().sum())
missing = clientes.isnull().mean().sort_values(ascending=False)
print("\nPorcentaje de nulos (top 10):")
print((missing*100).round(1))


Duplicados: 0

Porcentaje de nulos (top 10):
TotalPagosEfectuadosLocalmentePrevius     26.6
TotalPagosEfectuadosGlobalmentePrevius    26.6
DiasMaximosMoraCreditosGenerados          25.8
UsoAppWeb                                 24.5
FechaPrimerUso                            24.4
FechaUltimoUso                            24.4
CategoriaPrincipalCredito                 22.3
NumeroCreditosGCanalFPrevius              21.7
NumeroCreditosLEstadoActivosPrevius       21.7
NumeroCreditosGEstadoActivosPrevius       21.7
NumeroCreditosGPrevius                    21.7
NumeroCreditosLEstadoPagadosPrevius       21.7
NumeroCreditosLPrevius                    21.7
NumeroCreditosGCanalVPrevius              21.7
NumeroCreditosGEstadoPagadosPrevius       21.7
CodigoMunicipioEntregaTC                   1.2
TipoMunicipioEntregaTC                     1.2
FechaVinculacionCliente                    1.2
Edad                                       1.1
CupoAprobado                               0.4
NumeroIntentosF

# Correlaciones básicas con la variable objetivo PerdidaCartera

In [14]:
# Por si UsabilidadCupo está como texto
if clientes['UsabilidadCupo'].dtype == object:
    clientes['UsabilidadCupo'] = pd.to_numeric(clientes['UsabilidadCupo'], errors='coerce')

target = 'PerdidaCartera'

summary_rows = []
for col in clientes.columns:
    if col == target:
        continue
    nuniques = clientes[col].nunique()
    pct_null = clientes[col].isnull().mean()*100
    dtype = clientes[col].dtype
    top_values = clientes[col].value_counts(dropna=False).head(5).to_dict()
    summary_rows.append({
        'columna': col,
        'dtype': dtype,
        '%nulos': round(pct_null,1),
        'n_unicos': nuniques,
        'top5_valores': top_values
    })

summary_df = pd.DataFrame(summary_rows)
print("\n=== Resumen de columnas ===")
print(summary_df.to_string())

# Correlación numéricas con target
num_cols = clientes.select_dtypes(include=np.number).columns.tolist()
if target in num_cols:
    num_cols.remove(target)

corrs = clientes[num_cols].corrwith(clientes[target])
print("\n=== Correlación de variables numéricas con PerdidaCartera ===")
print(corrs.sort_values(ascending=False))

# Tablas cruzadas para categóricas con pocas categorías
cat_cols = clientes.select_dtypes(exclude=np.number).columns.tolist()
for col in cat_cols:
    if clientes[col].nunique() <= 20:  # ajusta el umbral si quieres
        print(f"\n=== Distribución {col} vs {target} ===")
        print(pd.crosstab(clientes[col], clientes[target], normalize='index').round(2))



=== Resumen de columnas ===
                                   columna           dtype  %nulos  n_unicos                                                                                                                                                 top5_valores
0                     IdentificadorCliente           int64     0.0    146939                                                                                                               {1: 1, 97963: 1, 97957: 1, 97958: 1, 97959: 1}
1                              FechaEvento          object     0.0    146939  {'2022-09-19T13:25:31.867Z': 1, '2023-10-17T23:18:34.553Z': 1, '2022-08-01T12:17:52.580Z': 1, '2023-08-05T17:49:48.307Z': 1, '2022-07-02T08:54:51.307Z': 1}
2                           UsabilidadCupo         float64     8.7    107704                                                                                          {nan: 12751, 1e+19: 2831, 0.1855: 1644, 0.5565: 1404, 0.27825: 900}
3                CategoriaPrincipal

## 📌 1. Variables con más poder predictivo inmediato



**DiasMora** tiene correlación altísima con PerdidaCartera (0.83), pero ojo:

Es una variable de **estado actual**. No se puede usar como predictor porque sería fuga.
Úsala solo para definir la etiqueta (PerdidaCartera) o para análisis descriptivo.

**Historial de créditos previos:**

- NumeroCreditosGEstadoActivosPrevius (0.43),

- NumeroCreditosLEstadoActivosPrevius (0.23),

- NumeroCreditosLPrevius (0.20),

- NumeroCreditosLEstadoPagadosPrevius (0.19),

- TotalPagosEfectuadosLocalmentePrevius (0.17).

→ Esto muestra que **tener más créditos activos previos** en general se asocia con mayor probabilidad de pérdida.
Buenas features para un modelo.

**Variables de fecha** (FechaVinculacionCliente, FechaPrimerUso) correlan débilmente, pero se pueden convertir a **antigüedad** (“meses desde vinculación”, “meses desde primer uso”). Eso suele mejorar la señal.

**Variables demográficas y cupo** (Edad, CupoAprobado, ScoreCrediticio) tienen correlación negativa, moderada: más edad, más cupo y más score → menos pérdida.

**Categorías de entrega y canal:**

- TipoMunicipioEntregaTC: clientes **VIRTUAL** tienen 28% de pérdida vs 16–20% en físico.

- CanalMunicipioEntregaTC: **Virtual** 28% vs **Físico** 20%.

- UsoAppWeb: App y Web similar (23–24%); puede servir combinado con otros factores.

**Genero:** casi balanceado F/M, con ligera diferencia (20% vs 25% pérdida).

## 📌 2. Variables a transformar antes de modelar

| Variable                                                        | Transformación recomendada                                                                                                                                                             |
| --------------------------------------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `FechaEvento`                                                   | No usar directo; se puede derivar “mes de compra” para análisis temporal, no para predicción del mismo evento.                                                                         |
| `UsabilidadCupo`                                                | Ya numérica, imputar `nan` con 0 o mediana; quizá binning.                                                                                                                             |
| `FechaVinculacionCliente` / `FechaPrimerUso` / `FechaUltimoUso` | Son números de Excel; convertir a datetime (`pd.to_datetime(..., origin='1899-12-30', unit='D')`) y luego calcular “antigüedad en meses” = (FechaEvento – FechaVinculacionCliente)/30. |
| `CodigoAlmacenEntregaTC` / `CodigoMunicipioEntregaTC`           | Demasiados valores; agrupar por top-N o por tipo (usar `TipoMunicipioEntregaTC` y `CanalMunicipioEntregaTC` en su lugar).                                                              |
| `Genero`                                                        | Reemplazar el 27 erróneo → “Desconocido”.                                                                                                                                              |


# 1904

In [9]:
# 1) Parsear fechas

# FechaEvento viene en ISO con 'Z' (UTC). La hacemos tz-aware y luego le quitamos la tz
clientes['FechaEvento_dt'] = (
    pd.to_datetime(clientes['FechaEvento'], errors='coerce', utc=True)
      .dt.tz_convert(None)  # elimina la tz y deja timestamps "naive"
)

# Fechas que vienen como número de Excel:
# - Vinculación y ÚltimoUso parecen en base 1900 (1899-12-30)
# - PrimerUso estaba en base 1904 (1904-01-01)
clientes['FechaVinculacionCliente_dt'] = pd.to_datetime(
    clientes['FechaVinculacionCliente'], errors='coerce',
    origin='1899-12-30', unit='D'
)
clientes['FechaUltimoUso_dt'] = pd.to_datetime(
    clientes['FechaUltimoUso'], errors='coerce',
    origin='1899-12-30', unit='D'
)
clientes['FechaPrimerUso_dt'] = pd.to_datetime(
    clientes['FechaPrimerUso'], errors='coerce',
    origin='1904-01-01', unit='D'   # << usa base 1904 para PrimerUso
)

# 2) Comprobar que todas sean tz-naive
print("FechaEvento_dt tz:", getattr(clientes['FechaEvento_dt'].dt, 'tz', None))
print("FechaVinculacionCliente_dt tz:", getattr(clientes['FechaVinculacionCliente_dt'].dt, 'tz', None))

# 3) Flag y corrección conservadora para PrimerUso
clientes['Flag_PrimerUsoAntesVinc'] = (
    (clientes['FechaPrimerUso_dt'] < clientes['FechaVinculacionCliente_dt']).astype(int)
)
mask_bad = clientes['FechaPrimerUso_dt'].isna() | \
           (clientes['FechaPrimerUso_dt'] < clientes['FechaVinculacionCliente_dt'])
clientes.loc[mask_bad,'FechaPrimerUso_corr'] = clientes.loc[mask_bad,'FechaVinculacionCliente_dt']
clientes.loc[~mask_bad,'FechaPrimerUso_corr'] = clientes.loc[~mask_bad,'FechaPrimerUso_dt']

# 4) Funciones para diferencias seguras
def safe_months(a, b):
    """(a - b) en meses, con NaN→0 y negativos recortados a 0."""
    delta_days = (a - b).dt.days
    delta_days = delta_days.where(delta_days.notna(), 0)
    delta_days = np.where(delta_days < 0, 0, delta_days)
    return delta_days / 30.0

def safe_days(a, b):
    delta_days = (a - b).dt.days
    delta_days = delta_days.where(delta_days.notna(), 0)
    delta_days = np.where(delta_days < 0, 0, delta_days)
    return delta_days

# 5) Calcular antigüedades ya corregidas
clientes['MesesDesdeVinculacion'] = safe_months(
    clientes['FechaEvento_dt'], clientes['FechaVinculacionCliente_dt']
)
clientes['MesesDesdePrimerUso'] = safe_months(
    clientes['FechaEvento_dt'], clientes['FechaPrimerUso_corr']
)
clientes['DiasDesdeUltimoUso'] = safe_days(
    clientes['FechaEvento_dt'], clientes['FechaUltimoUso_dt']
)

# 6) Verificar resultados
print(clientes[['FechaEvento_dt','FechaVinculacionCliente_dt','FechaPrimerUso_corr','FechaUltimoUso_dt','Flag_PrimerUsoAntesVinc']].head())
print(clientes[['MesesDesdeVinculacion','MesesDesdePrimerUso','DiasDesdeUltimoUso']].describe())


FechaEvento_dt tz: None
FechaVinculacionCliente_dt tz: None
           FechaEvento_dt FechaVinculacionCliente_dt FechaPrimerUso_corr  \
0 2022-09-19 13:25:31.867                 2019-09-11          2019-09-11   
1 2023-08-23 11:33:46.417                 2023-08-19          2023-08-19   
2 2022-10-01 14:59:48.920                 2019-03-21          2019-03-21   
3 2022-09-22 21:25:09.187                 2022-05-07          2022-05-07   
4 2023-03-19 17:48:52.310                 2021-10-27          2025-10-28   

  FechaUltimoUso_dt  Flag_PrimerUsoAntesVinc  
0        2022-07-23                        1  
1        2023-08-20                        1  
2        2022-09-26                        1  
3        2022-06-04                        1  
4        2022-09-03                        0  
       MesesDesdeVinculacion  MesesDesdePrimerUso  DiasDesdeUltimoUso
count          146939.000000        146939.000000       146939.000000
mean               16.894246             3.539166           9

In [10]:
# 1) ¿Cuántos casos tienen PrimerUso < Vinculación?
mask_incons = clientes['FechaPrimerUso_dt'] < clientes['FechaVinculacionCliente_dt']
print("Inconsistencias PrimerUso<Vinculación:", mask_incons.sum(), "de", len(clientes))

# 2) Magnitud (en días) de las inconsistencias
delta_neg = (clientes.loc[mask_incons, 'FechaPrimerUso_dt'] - clientes.loc[mask_incons, 'FechaVinculacionCliente_dt']).dt.days
print(delta_neg.describe())

# 3) Muestra 5 filas para inspección
cols_show = ['IdentificadorCliente','FechaVinculacionCliente_dt','FechaPrimerUso_dt','FechaEvento_dt']
print(clientes.loc[mask_incons, cols_show].head(5).to_string())


Inconsistencias PrimerUso<Vinculación: 4215 de 146939
count    4215.000000
mean     -680.971293
std      1005.981400
min     -7051.000000
25%      -694.000000
50%      -364.000000
75%      -162.000000
max        -1.000000
dtype: float64
   IdentificadorCliente FechaVinculacionCliente_dt FechaPrimerUso_dt          FechaEvento_dt
0                     1                 2019-09-11        2010-12-15 2022-09-19 13:25:31.867
1                     2                 2023-08-19        2012-07-22 2023-08-23 11:33:46.417
2                     3                 2019-03-21        2012-04-04 2022-10-01 14:59:48.920
3                     4                 2022-05-07        2013-07-26 2022-09-22 21:25:09.187
6                     7                 2022-08-19        2010-12-25 2022-12-04 21:33:38.130


Probar si es base 1904
Si el número de inconsistencias baja muchísimo, significa que esa columna venía en base 1904 y debes re-parsearla

In [ ]:

TARGET = 'PerdidaCartera'
# 2) Genero: arreglar valor erróneo 27 -> 'Desconocido', y ver distribución vs target
clientes['Genero'] = clientes['Genero'].replace({27: 'Desconocido'})
clientes['Genero'] = clientes['Genero'].fillna('Desconocido')

print("\n=== Distribución Genero ===")
print(clientes['Genero'].value_counts(dropna=False).to_string())

print("\n=== Genero vs PerdidaCartera (proporciones por fila) ===")
print(pd.crosstab(clientes['Genero'], clientes[TARGET], normalize='index').round(3).to_string())

# 3) Alta cardinalidad: Top-N + 'Otros' para ALMACÉN y MUNICIPIO (para EDA; NO meter al baseline)
def top_n_with_others(s, n=20):
    top = s.value_counts().nlargest(n).index
    return np.where(s.isin(top), s, 'Otros')

# Nota: CodigoAlmacenEntregaTC puede venir mezclado (texto '0xf' y números). Convirtamos a str para contar.
clientes['CodigoAlmacenEntregaTC_str'] = clientes['CodigoAlmacenEntregaTC'].astype(str)
clientes['AlmacenTop20'] = top_n_with_others(clientes['CodigoAlmacenEntregaTC_str'], n=20)

print("\n=== Top 10 CodigoAlmacenEntregaTC (como str) ===")
print(clientes['CodigoAlmacenEntregaTC_str'].value_counts().head(10).to_string())

print("\n=== AlmacenTop20 vs PerdidaCartera (primeras 10 filas) ===")
print(pd.crosstab(clientes['AlmacenTop20'], clientes[TARGET], normalize='index')
        .round(3).head(10).to_string())

# Municipio: tiene valor -1 para virtual; mantenemos el numérico pero también hacemos Top-N para EDA
# Si quieres preservar el -1 explícito como categoría:
clientes['MunicipioCat'] = clientes['CodigoMunicipioEntregaTC'].fillna(-999).astype(int).astype(str)
clientes['MunicipioTop20'] = top_n_with_others(clientes['MunicipioCat'], n=20)

print("\n=== Top 10 CodigoMunicipioEntregaTC ===")
print(clientes['MunicipioCat'].value_counts().head(10).to_string())

print("\n=== MunicipioTop20 vs PerdidaCartera (primeras 10 filas) ===")
print(pd.crosstab(clientes['MunicipioTop20'], clientes[TARGET], normalize='index')
        .round(3).head(10).to_string())

# 4) Limpiar ruido en TipoMunicipioEntregaTC y analizar
#    (arreglamos el encoding 'PEQUEÑO' -> 'PEQUEÑO')
clientes['TipoMunicipioEntregaTC'] = (
    clientes['TipoMunicipioEntregaTC']
      .replace({'PEQUEÑO': 'PEQUEÑO'})
      .fillna('Desconocido')
)

print("\n=== TipoMunicipioEntregaTC (limpio) ===")
print(clientes['TipoMunicipioEntregaTC'].value_counts().to_string())

print("\n=== TipoMunicipioEntregaTC vs PerdidaCartera ===")
print(pd.crosstab(clientes['TipoMunicipioEntregaTC'], clientes[TARGET], normalize='index')
        .round(3).to_string())

# 5) CanalMunicipioEntregaTC: completar nulos y revisar
clientes['CanalMunicipioEntregaTC'] = clientes['CanalMunicipioEntregaTC'].fillna('Desconocido')
print("\n=== CanalMunicipioEntregaTC vs PerdidaCartera ===")
print(pd.crosstab(clientes['CanalMunicipioEntregaTC'], clientes[TARGET], normalize='index')
        .round(3).to_string())

# >>> RECORDATORIO PARA EL MODELO BASELINE <<<
# - NO usar: 'CodigoAlmacenEntregaTC_str', 'AlmacenTop20', 'MunicipioCat', 'MunicipioTop20'
#   (son útiles para EDA, pero por ahora los excluimos del baseline por estacionalidad y alta cardinalidad).
# - SÍ usar: 'TipoMunicipioEntregaTC' y 'CanalMunicipioEntregaTC' (categóricas manejables),
#            'Genero' (ya limpiado), 'UsabilidadCupo' (num) y su bin si te interesa,
#            'MesesDesdeVinculacion', 'MesesDesdePrimerUso', 'DiasDesdeUltimoUso',
#            todo el bloque de históricos G/L (créditos activos/pagados, totales, etc.),
#            'Flag_PrimerUsoTemu', 'Flag_PrimerUsoAntesVinc'.


=== Distribución Genero ===
Genero
Masculino      70506
Femenino       65494
Desconocido    10939

=== Genero vs PerdidaCartera (proporciones por fila) ===
PerdidaCartera      0      1
Genero                      
Desconocido     0.756  0.244
Femenino        0.801  0.199
Masculino       0.753  0.247

=== Top 10 CodigoAlmacenEntregaTC (como str) ===
CodigoAlmacenEntregaTC_str
0xf     1824
11        16
34        16
3818      15
1298      15
7126      15
270       14
3211      14
4169      14
3336      13

=== AlmacenTop20 vs PerdidaCartera (primeras 10 filas) ===
PerdidaCartera      0      1
AlmacenTop20                
0xf             0.940  0.060
1               0.923  0.077
11              0.750  0.250
11718           0.846  0.154
1298            0.600  0.400
270             0.857  0.143
3211            0.714  0.286
3336            0.923  0.077
34              1.000  0.000
3519            0.769  0.231

=== Top 10 CodigoMunicipioEntregaTC ===
MunicipioCat
-1      54116
130     18957
2

**CodigoAlmacenEntregaTC / CodigoMunicipioEntregaTC**

– Alta cardinalidad: no los entra crudos al baseline.
– En su lugar TipoMunicipioEntregaTC y CanalMunicipioEntregaTC, que ya vimos que tienen diferencias claras en tasa de pérdida (por ejemplo VIRTUAL 27 % vs Físico 19 %).
– Más adelante quieres refinar, agrupar los top-N más significativos (como hicimos en AlmacenTop20 y MunicipioTop20) y meterlos como categorías adicionales, pero para arrancar mejor simplificar.

**Genero**

– Ya el 27 → 'Desconocido'.
– Esta variable la puedes dejar tal cual para OneHot en el baseline (Masculino, Femenino, Desconocido).

**UsabilidadCupo**

– Ya está numérica y con imputación por mediana.
– columna continua UsabilidadCupo para el modelo lineal y/o LightGBM.
– también UsabilidadCupo_bin para análisis interpretativo (en un modelo lineal es más fácil ver el efecto por segmento).

**Antigüedad**

– Ya MesesDesdeVinculacion y MesesDesdePrimerUso calculadas y corregidas.
– Usa directamente como numéricas.
– Guarda también Flag_PrimerUsoAntesVinc y Flag_PrimerUsoTemu (primera compra en Temu) como features binarias.

Qué ya está bien

Fechas: parseadas y antigüedades creadas (MesesDesdeVinculacion, MesesDesdePrimerUso), con flag de inconsistencia Flag_PrimerUsoAntesVinc. ✅

Categóricas claves con señal: TipoMunicipioEntregaTC (VIRTUAL 28% vs 16–20%), CanalMunicipioEntregaTC (Virtual 28% vs Físico 20%), Genero. ✅

Alta cardinalidad (CodigoAlmacenEntregaTC, CodigoMunicipioEntregaTC) dejadas fuera del baseline. ✅

Gaps y ajustes recomendados

UsabilidadCupo: tienes valores absurdos (1e19), y NaN (~8.7%).

Acción: convertir a numérico, cap (winsorizar) al rango [0, 2] (o percentil 99), crear flags de outlier y de nulo, y rellenar con mediana.

Pagos/Créditos previos: muchas columnas con ~22–27% NaN. Es muy probable que “NaN ≈ no historial / no aplica”.

Acción: imputar 0 + flag de nulo por columna (para que el modelo capte la diferencia entre “0 real” y “desconocido”).

ScoreCrediticio: 0 significa “sin información”, no “mal score”.

Acción: crear ScoreSinInfo = (ScoreCrediticio==0) y dejar ScoreCrediticio tal cual (o poner NaN cuando es 0 y luego imputar/median). (Yo prefiero flag + valor sin tocar).

CupoAprobado: muy sesgado (3e9–3e10).

Acción: crear log_CupoAprobado = log1p(CupoAprobado); conservar el original.

Edad: validar rango plausible (>=18 y <=100); fuera de rango → NaN (+ flag si quieres).

Leakage de fechas: FechaUltimoUso puede ser posterior al evento en algunos casos (p. ej., compra Temu hoy, “último uso” mañana) → no queremos fuga.

Acción: ya calculaste DiasDesdeUltimoUso con clip a 0; añade Flag_UltimoUsoPosterior = 1 si FechaUltimoUso_dt > FechaEvento_dt.

Categorías nulas/raras:

UsoAppWeb (24.5% NaN) → usar “Desconocido”.

CategoriaPrincipalCredito (22.3% NaN) → usar “Desconocido”.

CanalMunicipioEntregaTC="Desconocido" (16 filas, 69% pérdida) → mantener como categoría; el regularizador lo controla.

## 📌 3. Feature engineering sugerido

**Bandera “primer uso en Temu”:** first_use_temu = (NumeroCreditosGPrevius==0).astype(int).

**Ratios:**

- ratio_pagos_local_global = TotalPagosEfectuadosLocalmentePrevius / (TotalPagosEfectuadosGlobalmentePrevius+1)

- creditos_activos_ratio = NumeroCreditosGEstadoActivosPrevius/(NumeroCreditosGPrevius+1)

**Antigüedad:** meses desde FechaVinculacionCliente a FechaEvento.

**Bandera canal virtual:** de TipoMunicipioEntregaTC o CanalMunicipioEntregaTC.

**Score crediticio bucket:** 0 (sin info), bajo, medio, alto.

Estas transformaciones dan un set de features más estable y explicable.

In [ ]:
# ==========================================
# LIMPIEZA + FEATURE ENGINEERING (CONSISTENTE)
# ==========================================

TARGET = 'PerdidaCartera'
df = clientes  # alias corto si quieres

# -------------------------------
# 0) SANITY CHECKS PREVIOS
# -------------------------------
# Asegurar columnas clave existentes
required_cols = [
    'UsabilidadCupo','FechaUltimoUso_dt','FechaEvento_dt',
    'TotalPagosEfectuadosLocalmentePrevius','TotalPagosEfectuadosGlobalmentePrevius',
    'NumeroCreditosGEstadoActivosPrevius','NumeroCreditosGPrevius',
    'MesesDesdeVinculacion','MesesDesdePrimerUso','DiasDesdeUltimoUso',
    'CanalMunicipioEntregaTC','TipoMunicipioEntregaTC','ScoreCrediticio',
    'CategoriaPrincipalCredito','UsoAppWeb','Genero','Edad','CupoAprobado',
    'NumeroCreditosGCanalFPrevius','NumeroCreditosGCanalVPrevius',
    'NumeroCreditosGEstadoPagadosPrevius','NumeroCreditosLPrevius',
    'NumeroCreditosLEstadoActivosPrevius','NumeroCreditosLEstadoPagadosPrevius'
]
missing_req = [c for c in required_cols if c not in df.columns]
if missing_req:
    print("ADVERTENCIA: faltan columnas esperadas:", missing_req)

# -------------------------------
# 1) USABILIDAD CUPO (numérico + caps + flags + imputación)
# -------------------------------
df['UsabilidadCupo'] = pd.to_numeric(df['UsabilidadCupo'], errors='coerce')
df['Flag_Usab_NaN']   = df['UsabilidadCupo'].isna().astype(int)
# Cap por negocio: ratio esperado [0,1], toleramos hasta 2
df['Flag_Usab_Outlier'] = ((df['UsabilidadCupo'] < 0) | (df['UsabilidadCupo'] > 2)).astype(int)
df.loc[df['UsabilidadCupo'] < 0, 'UsabilidadCupo'] = np.nan
df.loc[df['UsabilidadCupo'] > 2, 'UsabilidadCupo'] = np.nan
# Imputación mediana (robusta)
df['UsabilidadCupo'] = df['UsabilidadCupo'].fillna(df['UsabilidadCupo'].median())

print("\n=== Auditoría UsabilidadCupo ===")
print(df[['UsabilidadCupo','Flag_Usab_NaN','Flag_Usab_Outlier']].describe(include='all').to_string())

# -------------------------------
# 2) FLAGS DE FECHAS (LEAKAGE)
# -------------------------------
# Último uso posterior al evento → posible fuga si se usara cruda; lo marcamos
df['Flag_UltimoUsoPosterior'] = (df['FechaUltimoUso_dt'] > df['FechaEvento_dt']).fillna(False).astype(int)
print("\n=== Flag_UltimoUsoPosterior (cuenta) ===")
print(df['Flag_UltimoUsoPosterior'].value_counts(dropna=False).to_string())

# -------------------------------
# 3) NUMÉRICAS CON NULOS: imputar y flaggear
# -------------------------------
num_cols_candidates = [
    'DiasMaximosMoraCreditosGenerados',
    'NumeroCreditosGPrevius','NumeroCreditosGCanalFPrevius','NumeroCreditosGCanalVPrevius',
    'NumeroCreditosGEstadoActivosPrevius','NumeroCreditosGEstadoPagadosPrevius',
    'NumeroCreditosLPrevius','NumeroCreditosLEstadoActivosPrevius','NumeroCreditosLEstadoPagadosPrevius',
    'TotalPagosEfectuadosGlobalmentePrevius','TotalPagosEfectuadosLocalmentePrevius',
    'NumeroIntentosFallidos','CupoAprobado','ScoreCrediticio','Edad',
    'MesesDesdeVinculacion','MesesDesdePrimerUso','DiasDesdeUltimoUso'
]

for c in num_cols_candidates:
    if c in df.columns:
        df[f'Flag_{c}_NaN'] = df[c].isna().astype(int)
        # Conteos/pagos/historial: NaN ≈ sin historial → 0
        if c.startswith('NumeroCreditos') or c.startswith('TotalPagos'):
            df[c] = df[c].fillna(0)
        elif c in ['DiasMaximosMoraCreditosGenerados','Edad','NumeroIntentosFallidos']:
            df[c] = df[c].fillna(0)
        # Cupo/Score se tratan abajo

# -------------------------------
# 4) SCORE & CUPO
# -------------------------------
# Score: 0 => sin información (no imputar, es señal)
df['ScoreSinInfo'] = (df['ScoreCrediticio'] == 0).astype(int)
# Si hay NaN, imputar con mediana de los >0 (mantiene la interpretación del 0)
if df['ScoreCrediticio'].isna().any():
    med_pos = df.loc[df['ScoreCrediticio']>0, 'ScoreCrediticio'].median()
    df['ScoreCrediticio'] = df['ScoreCrediticio'].fillna(med_pos)

# Cupo: escala log (robusta a colas largas)
df['log_CupoAprobado'] = np.log1p(df['CupoAprobado'])
if df['log_CupoAprobado'].isna().any():
    df['log_CupoAprobado'] = df['log_CupoAprobado'].fillna(df['log_CupoAprobado'].median())

print("\n=== Auditoría Score/Cupo ===")
print(df[['ScoreCrediticio','ScoreSinInfo','CupoAprobado','log_CupoAprobado']].describe().to_string())

# -------------------------------
# 5) CATEGÓRICAS: completar nulos y limpiar
# -------------------------------
for c in ['CategoriaPrincipalCredito','UsoAppWeb','Genero','TipoMunicipioEntregaTC','CanalMunicipioEntregaTC']:
    if c in df.columns:
        df[c] = df[c].fillna('Desconocido')

# Genero ya corregido previamente (27 -> 'Desconocido'); si no, se podría asegurar aquí:
df['Genero'] = df['Genero'].replace({27:'Desconocido'})

# -------------------------------
# 6) CONTROL DE RANGOS
# -------------------------------
# Edad plausible
df['Flag_Edad_Out'] = (~df['Edad'].between(18, 100, inclusive='both')).fillna(False).astype(int)
df.loc[df['Flag_Edad_Out']==1, 'Edad'] = np.nan
df['Edad'] = df['Edad'].fillna(df['Edad'].median())

# -------------------------------
# 7) FEATURES DERIVADAS CLAVE
# -------------------------------
# 7.1 Bandera “primer uso en Temu”
if 'Flag_PrimerUsoTemu' not in df.columns and 'NumeroCreditosGPrevius' in df.columns:
    df['Flag_PrimerUsoTemu'] = (df['NumeroCreditosGPrevius'] == 0).astype(int)

print("\n=== Flag_PrimerUsoTemu (0/1, conteo) ===")
print(df['Flag_PrimerUsoTemu'].value_counts(dropna=False).to_string())

# 7.2 Ratios
df['ratio_pagos_local_global'] = (
    df['TotalPagosEfectuadosLocalmentePrevius'].fillna(0) /
    (df['TotalPagosEfectuadosGlobalmentePrevius'].fillna(0) + 1.0)
)
df['creditos_activos_ratio'] = (
    df['NumeroCreditosGEstadoActivosPrevius'].fillna(0) /
    (df['NumeroCreditosGPrevius'].fillna(0) + 1.0)
)
# Limpiar ratios (NaN/inf→0) y cap en [0,1]
for c in ['ratio_pagos_local_global','creditos_activos_ratio']:
    df[c] = df[c].replace([np.inf, -np.inf], np.nan).fillna(0).clip(lower=0, upper=1)

print("\n=== Resumen ratios ===")
print(df[['ratio_pagos_local_global','creditos_activos_ratio']].describe().to_string())

# 7.3 Bandera de canal virtual (si cualquiera lo indica)
df['Flag_CanalVirtual'] = (
    (df['CanalMunicipioEntregaTC'].astype(str).str.lower() == 'virtual') |
    (df['TipoMunicipioEntregaTC'].astype(str).str.upper() == 'VIRTUAL')
).astype(int)

print("\n=== Flag_CanalVirtual (0/1, conteo) ===")
print(df['Flag_CanalVirtual'].value_counts(dropna=False).to_string())
print("\n=== Flag_CanalVirtual vs PerdidaCartera ===")
print(pd.crosstab(df['Flag_CanalVirtual'], df[TARGET], normalize='index').round(3).to_string())

# 7.4 Buckets de ScoreCrediticio (explicable)
df['ScoreBucket'] = 'sin_info'
mask_pos = df['ScoreCrediticio'] > 0
if mask_pos.sum() > 0:
    b1, b2 = df.loc[mask_pos, 'ScoreCrediticio'].quantile([0.33, 0.66]).values
    df.loc[mask_pos & (df['ScoreCrediticio'] <= b1), 'ScoreBucket'] = 'bajo'
    df.loc[mask_pos & (df['ScoreCrediticio'] >  b1) & (df['ScoreCrediticio'] <= b2), 'ScoreBucket'] = 'medio'
    df.loc[mask_pos & (df['ScoreCrediticio'] >  b2), 'ScoreBucket'] = 'alto'
df['ScoreBucket'] = pd.Categorical(df['ScoreBucket'], categories=['sin_info','bajo','medio','alto'], ordered=True)

print("\n=== ScoreBucket (conteo) ===")
print(df['ScoreBucket'].value_counts(dropna=False).to_string())
print("\n=== ScoreBucket vs PerdidaCartera ===")
print(pd.crosstab(df['ScoreBucket'], df[TARGET], normalize='index').round(3).to_string())

# 7.5 Antigüedad (ya creada)
print("\n=== Antigüedad (MesesDesdeVinculacion) — resumen ===")
print(df['MesesDesdeVinculacion'].describe().to_string())

# -------------------------------
# 8) LISTA FINAL DE FEATURES PARA BASELINE
# -------------------------------
drop_from_features = [
    'IdentificadorCliente','FechaEvento','FechaVinculacionCliente','FechaPrimerUso','FechaUltimoUso',
    'FechaEvento_dt','FechaVinculacionCliente_dt','FechaPrimerUso_dt','FechaUltimoUso_dt','FechaPrimerUso_corr',
    'CodigoAlmacenEntregaTC','CodigoAlmacenEntregaTC_str','AlmacenTop20',
    'CodigoMunicipioEntregaTC','MunicipioCat','MunicipioTop20','MesCompra',
    'DiasMora'  # fuga
]

num_final = [
    'UsabilidadCupo','MesesDesdeVinculacion','MesesDesdePrimerUso','DiasDesdeUltimoUso',
    'NumeroCreditosGPrevius','NumeroCreditosGCanalFPrevius','NumeroCreditosGCanalVPrevius',
    'NumeroCreditosGEstadoActivosPrevius','NumeroCreditosGEstadoPagadosPrevius',
    'NumeroCreditosLPrevius','NumeroCreditosLEstadoActivosPrevius','NumeroCreditosLEstadoPagadosPrevius',
    'TotalPagosEfectuadosGlobalmentePrevius','TotalPagosEfectuadosLocalmentePrevius',
    'NumeroIntentosFallidos','ScoreCrediticio','ScoreSinInfo','CupoAprobado','log_CupoAprobado','Edad',
    'Flag_Usab_NaN','Flag_Usab_Outlier','Flag_PrimerUsoAntesVinc','Flag_UltimoUsoPosterior','Flag_Edad_Out',
    # Nuevas numéricas:
    'ratio_pagos_local_global','creditos_activos_ratio'
] + [f'Flag_{c}_NaN' for c in num_cols_candidates if f'Flag_{c}_NaN' in df.columns]

cat_final = [
    'Genero','TipoMunicipioEntregaTC','CanalMunicipioEntregaTC','UsoAppWeb','CategoriaPrincipalCredito',
    'Flag_PrimerUsoTemu',   # puedes moverla a numéricas si prefieres (0/1)
    'ScoreBucket'           # bucket explicable del score
]

# Filtrar por existencia y excluir caídas
num_final = [c for c in num_final if c in df.columns and c not in drop_from_features]
cat_final = [c for c in cat_final if c in df.columns and c not in drop_from_features]

X = df[num_final + cat_final].copy()
y = df[TARGET].copy()

print("\n=== Resumen columnas finales ===")
print("Numéricas (", len(num_final), "):", num_final[:12], "...")
print("Categóricas (", len(cat_final), "):", cat_final)

# -------------------------------
# 9) AUDITORÍAS POST-LIMPIEZA
# -------------------------------
print("\n=== Correlación numéricas (post-limpieza) con PerdidaCartera ===")
print(X[num_final].corrwith(y).sort_values(ascending=False).to_string())

for c in cat_final:
    if X[c].nunique() <= 40:
        print(f"\n=== {c} vs {TARGET} ===")
        print(pd.crosstab(X[c], y, normalize='index').round(3).to_string())

print("\n=== Balance de clases ===")
print(y.value_counts(normalize=True).rename({0:'% no pérdida',1:'% pérdida'}).round(3).to_string())

# Limpieza de columnas auxiliares de EDA (si quedaron)
for c in ['ratio_pagos_local_global_bin','creditos_activos_ratio_bin']:
    if c in df.columns:
        df.drop(columns=[c], inplace=True, errors='ignore')

# Deja el resultado consolidado en 'X' y 'y' para el baseline.
clientes = df  # si quieres mantener 'clientes' como DataFrame resultante



=== Auditoría UsabilidadCupo ===
       UsabilidadCupo  Flag_Usab_NaN  Flag_Usab_Outlier
count   146939.000000       146939.0           146939.0
mean         0.351514            0.0                0.0
std          0.256366            0.0                0.0
min          0.002082            0.0                0.0
25%          0.166692            0.0                0.0
50%          0.278900            0.0                0.0
75%          0.498059            0.0                0.0
max          0.999995            0.0                0.0

=== Flag_UltimoUsoPosterior (cuenta) ===
Flag_UltimoUsoPosterior
0    146939

=== Auditoría Score/Cupo ===
       ScoreCrediticio   ScoreSinInfo  CupoAprobado  log_CupoAprobado
count    146939.000000  146939.000000  1.462790e+05     146939.000000
mean        530.002688       0.158569  1.395009e+10         22.934600
std         254.256067       0.365275  1.151370e+10          0.983373
min          -9.000000       0.000000  1.000000e+09         20.723266
25% 

Hagamos una auditoría de rangos y valores raros bien completa (solo prints/tablas, sin gráficos) para:

numéricas: min/max, p5/p95, conteo de negativos, ceros, IQR outliers, z-score outliers.

columnas “conteo” (créditos/pagos): chequear negativos y “no enteros”.

ratios: verificar que estén en [0,1].

categóricas: contar niveles raros (frecuencia < 0.1%) y top-10.

edad y cupo: rangos plausibles.

un resumen de “issues” por columna.

In [17]:
df = clientes.copy()

# =============== CONFIG ===============
# Columnas que conceptualmente son "conteos" (deben ser >=0 y casi enteras)
count_like = [
    'NumeroCreditosGPrevius','NumeroCreditosGCanalFPrevius','NumeroCreditosGCanalVPrevius',
    'NumeroCreditosGEstadoActivosPrevius','NumeroCreditosGEstadoPagadosPrevius',
    'NumeroCreditosLPrevius','NumeroCreditosLEstadoActivosPrevius','NumeroCreditosLEstadoPagadosPrevius',
    'TotalPagosEfectuadosGlobalmentePrevius','TotalPagosEfectuadosLocalmentePrevius',
    'NumeroIntentosFallidos'
]
# Ratios esperados en [0,1]
ratio_like = ['UsabilidadCupo','ratio_pagos_local_global','creditos_activos_ratio']

# Numéricas a auditar: usa las finales si existen, si no, saca numéricas del DF
if 'num_final' in globals():
    num_cols = [c for c in num_final if c in df.columns]
else:
    num_cols = df.select_dtypes(include=[np.number]).columns.tolist()

# Categóricas a auditar
if 'cat_final' in globals():
    cat_cols = [c for c in cat_final if c in df.columns]
else:
    cat_cols = df.select_dtypes(exclude=[np.number]).columns.tolist()

print(f"\n### Columnas numéricas a auditar ({len(num_cols)})")
print(num_cols)
print(f"\n### Columnas categóricas a auditar ({len(cat_cols)})")
print(cat_cols)

# =============== 1) AUDITORÍA NUMÉRICAS ===============
rows = []
for c in num_cols:
    s = pd.to_numeric(df[c], errors='coerce')
    n = s.notna().sum()
    if n == 0:
        rows.append([c, 'NA', 'NA', 'NA', 'NA', 'NA', 'NA', 0, 0, 0, 0, np.nan, np.nan])
        continue
    desc = s.describe(percentiles=[.01,.05,.25,.5,.75,.95,.99])
    q1, q3 = desc['25%'], desc['75%']
    iqr = q3 - q1
    # IQR outliers
    lower, upper = q1 - 1.5*iqr, q3 + 1.5*iqr
    iqr_out = ((s < lower) | (s > upper)).sum()
    # z-score outliers (solo si std>0)
    if desc['std'] > 0:
        z = (s - desc['mean'])/desc['std']
        z_out = (z.abs() > 4).sum()
    else:
        z_out = 0
    neg = (s < 0).sum()
    zeros = (s == 0).sum()
    infs = np.isinf(s).sum()
    nans = s.isna().sum()
    rows.append([
        c,
        float(desc['min']), float(desc['1%']), float(desc['5%']), float(desc['50%']),
        float(desc['95%']), float(desc['99%']),
        int(neg), int(zeros), int(iqr_out), int(z_out), int(infs), int(nans)
    ])

num_audit = pd.DataFrame(rows, columns=[
    'col','min','p01','p05','p50','p95','p99',
    'negativos','ceros','outliers_IQR','outliers_Z(|z|>4)','inf','nans'
]).sort_values(['outliers_IQR','outliers_Z(|z|>4)'], ascending=False)

print("\n=== Auditoría numéricas (rangos & outliers) ===")
print(num_audit.to_string(index=False))

# =============== 2) ESPECÍFICO: CONTEOS ===============
rows = []
for c in [x for x in count_like if x in df.columns]:
    s = pd.to_numeric(df[c], errors='coerce')
    non_int = ((s.dropna() % 1) != 0).sum()
    neg = (s < 0).sum()
    maxv = s.max()
    rows.append([c, int(neg), int(non_int), int(maxv)])
count_check = pd.DataFrame(rows, columns=['col','negativos','no_enteros','max']).sort_values('negativos', ascending=False)

print("\n=== Chequeo columnas tipo conteo (>=0, enteras) ===")
print(count_check.to_string(index=False))

# =============== 3) ESPECÍFICO: RATIOS [0,1] ===============
rows = []
for c in [x for x in ratio_like if x in df.columns]:
    s = pd.to_numeric(df[c], errors='coerce')
    below0 = (s < 0).sum()
    above1 = (s > 1).sum()
    rows.append([c, int(below0), int(above1), float(s.min()), float(s.max())])
ratio_check = pd.DataFrame(rows, columns=['col','<0','>1','min','max'])

print("\n=== Chequeo ratios esperados en [0,1] ===")
print(ratio_check.to_string(index=False))

# =============== 4) CATEGÓRICAS: RARE LEVELS ===============
def rare_levels_report(series, rare_frac=0.001):
    vc = series.value_counts(dropna=False)
    total = vc.sum()
    rare = vc[vc/total < rare_frac]
    return vc.head(10), rare

for c in cat_cols:
    top10, rare = rare_levels_report(df[c].astype(str))
    print(f"\n=== {c} — top10 ===")
    print(top10.to_string())
    if len(rare) > 0:
        print(f"\n{c} — niveles raros (<0.1% del total): {len(rare)}")
        print(rare.head(10).to_string())

# =============== 5) REGLAS DE SANIDAD ADICIONALES ===============
# Edad plausible (ya saneaste, pero reportemos rango final)
if 'Edad' in df.columns:
    print("\n=== Edad — rango tras saneo ===")
    print(df['Edad'].describe().to_string())

# CupoAprobado: mirar rango y extremos
if 'CupoAprobado' in df.columns:
    print("\n=== CupoAprobado — rango ===")
    print(df['CupoAprobado'].describe(percentiles=[.01,.05,.95,.99]).to_string())

# =============== 6) RESUMEN DE ISSUES (lista corta) ===============
issues = []

# numéricas con negativos o inf
for _, r in num_audit.iterrows():
    if r['negativos'] > 0 or r['inf'] > 0:
        issues.append(f"{r['col']}: negativos={int(r['negativos'])}, inf={int(r['inf'])}")

# conteos no enteros
for _, r in count_check.iterrows():
    if r['no_enteros'] > 0 or r['negativos'] > 0:
        issues.append(f"{r['col']}: no_enteros={int(r['no_enteros'])}, negativos={int(r['negativos'])}")

# ratios fuera de [0,1]
for _, r in ratio_check.iterrows():
    if r['<0'] > 0 or r['>1'] > 0:
        issues.append(f"{r['col']}: fuera_[0,1] (<0:{int(r['<0'])}, >1:{int(r['>1'])})")

print("\n=== RESUMEN ISSUES DETECTADOS ===")
if issues:
    for s in issues[:30]:
        print("-", s)
    if len(issues) > 30:
        print(f"... y {len(issues)-30} más")
else:
    print("No se detectaron problemas relevantes más allá de outliers típicos.")



### Columnas numéricas a auditar (45)
['UsabilidadCupo', 'MesesDesdeVinculacion', 'MesesDesdePrimerUso', 'DiasDesdeUltimoUso', 'NumeroCreditosGPrevius', 'NumeroCreditosGCanalFPrevius', 'NumeroCreditosGCanalVPrevius', 'NumeroCreditosGEstadoActivosPrevius', 'NumeroCreditosGEstadoPagadosPrevius', 'NumeroCreditosLPrevius', 'NumeroCreditosLEstadoActivosPrevius', 'NumeroCreditosLEstadoPagadosPrevius', 'TotalPagosEfectuadosGlobalmentePrevius', 'TotalPagosEfectuadosLocalmentePrevius', 'NumeroIntentosFallidos', 'ScoreCrediticio', 'ScoreSinInfo', 'CupoAprobado', 'log_CupoAprobado', 'Edad', 'Flag_Usab_NaN', 'Flag_Usab_Outlier', 'Flag_PrimerUsoAntesVinc', 'Flag_UltimoUsoPosterior', 'Flag_Edad_Out', 'ratio_pagos_local_global', 'creditos_activos_ratio', 'Flag_DiasMaximosMoraCreditosGenerados_NaN', 'Flag_NumeroCreditosGPrevius_NaN', 'Flag_NumeroCreditosGCanalFPrevius_NaN', 'Flag_NumeroCreditosGCanalVPrevius_NaN', 'Flag_NumeroCreditosGEstadoActivosPrevius_NaN', 'Flag_NumeroCreditosGEstadoPagadosPrevi

En este dataset específico, el diccionario dice explícitamente:
ScoreCrediticio <= 0 = “sin información suficiente” (no que sea “muy malo”).
Así que, para evitar introducir un sesgo que contradice la semántica del dataset, la codificación correcta por defecto es tratar <= 0 como “sin_info” (y por eso te propuse mapear negativos a 0, con un Flag_Score_Negativo por si hubiera señal residual).

In [19]:
df = clientes  # alias

# =========================
# 1) SCORE NEGATIVO -> 0
# =========================
df['Flag_Score_Negativo'] = (df['ScoreCrediticio'] < 0).astype(int)
df.loc[df['ScoreCrediticio'] < 0, 'ScoreCrediticio'] = 0

# Recalcular ScoreBucket (manteniendo regla 0 => sin_info)
df['ScoreBucket'] = 'sin_info'
mask_pos = df['ScoreCrediticio'] > 0
if mask_pos.sum() > 0:
    b1, b2 = df.loc[mask_pos, 'ScoreCrediticio'].quantile([0.33, 0.66]).values
    df.loc[mask_pos & (df['ScoreCrediticio'] <= b1), 'ScoreBucket'] = 'bajo'
    df.loc[mask_pos & (df['ScoreCrediticio'] >  b1) & (df['ScoreCrediticio'] <= b2), 'ScoreBucket'] = 'medio'
    df.loc[mask_pos & (df['ScoreCrediticio'] >  b2), 'ScoreBucket'] = 'alto'
df['ScoreBucket'] = pd.Categorical(df['ScoreBucket'], categories=['sin_info','bajo','medio','alto'], ordered=True)

print("\n>>> ScoreCrediticio: negativos forzados a 0")
print(df[['ScoreCrediticio','Flag_Score_Negativo']].describe().to_string())
print("\nScoreBucket (rehecho):")
print(df['ScoreBucket'].value_counts().to_string())

# ==============================================
# 2) WINSORIZACIÓN (cap p99) + FLAGS DE RECORTE
# ==============================================
def cap_with_flag(s, upper):
    flag = (s > upper).astype(int)
    return np.where(s > upper, upper, s), flag

# Columnas con colas largas (según tu auditoría)
cap_cols = [
    'DiasDesdeUltimoUso', 'MesesDesdeVinculacion',
    'TotalPagosEfectuadosGlobalmentePrevius', 'TotalPagosEfectuadosLocalmentePrevius',
    'NumeroCreditosGPrevius','NumeroCreditosGCanalFPrevius','NumeroCreditosGCanalVPrevius',
    'NumeroCreditosGEstadoActivosPrevius','NumeroCreditosGEstadoPagadosPrevius',
    'NumeroCreditosLPrevius','NumeroCreditosLEstadoActivosPrevius','NumeroCreditosLEstadoPagadosPrevius'
]
# Calcula p99 por columna y aplica cap
caps_info = {}
for c in cap_cols:
    if c in df.columns:
        p99 = df[c].quantile(0.99)
        capped, flag = cap_with_flag(df[c].fillna(0), p99)
        df[c] = capped
        df[f'Flag_{c}_Capped'] = flag
        caps_info[c] = float(p99)

print("\n>>> p99 usados para cap:")
print(pd.Series(caps_info).round(3).to_string())

# Verificación rápida post-cap (unas cuantas columnas)
check_cols = [c for c in cap_cols if c in df.columns][:8]
print("\nResumen post-cap (primeras columnas):")
print(df[check_cols].describe(percentiles=[.5,.95,.99]).to_string())

# ======================================
# 3) AGRUPAR CATEGORÍAS RARAS (<0.1%)
# ======================================
def lump_rare(series, min_frac=0.001, rare_label='OtrosRare'):
    vc = series.value_counts(dropna=False)
    cutoff = series.shape[0] * min_frac
    rare_levels = vc[vc < cutoff].index
    out = series.copy()
    out = out.where(~out.isin(rare_levels), rare_label)
    return out

if 'CategoriaPrincipalCredito' in df.columns:
    before_n = df['CategoriaPrincipalCredito'].nunique(dropna=False)
    df['CategoriaPrincipalCredito'] = lump_rare(df['CategoriaPrincipalCredito'].astype(str), min_frac=0.001)
    after_n  = df['CategoriaPrincipalCredito'].nunique(dropna=False)
    print(f"\n>>> CategoriaPrincipalCredito: niveles antes={before_n}, después={after_n} (raros → 'OtrosRare')")

# ======================================
# 4) ACTUALIZAR LISTAS DE FEATURES
# ======================================
# Asegurar que agregamos las nuevas flags de cap
new_cap_flags = [f'Flag_{c}_Capped' for c in cap_cols if f'Flag_{c}_Capped' in df.columns]

# num_final / cat_final ya existen; los extendemos si hace falta
extra_num = ['Flag_Score_Negativo'] + new_cap_flags
for c in extra_num:
    if c in df.columns and c not in num_final:
        num_final.append(c)

# la categórica editada ya está en cat_final; no cambia lista
# (pero advertimos que cambió su cardinalidad)
print("\n>>> Nuevas numéricas añadidas:")
print([c for c in extra_num if c in df.columns])

# ==========================
# 5) MINI AUDITORÍA FINAL
# ==========================
from sklearn.metrics import roc_auc_score

print("\n=== Chequeo ratios siguen en [0,1] ===")
for c in ['UsabilidadCupo','ratio_pagos_local_global','creditos_activos_ratio']:
    if c in df.columns:
        print(c, "min=", df[c].min(), "max=", df[c].max())

print("\n=== Resumen top de columnas cappeadas (flags=1) ===")
flag_rates = {f: float(df[f].mean()) for f in new_cap_flags}
print(pd.Series(flag_rates).sort_values(ascending=False).round(4).to_string())

# Re-armar X, y por si vas directo a entrenar
X = df[num_final + cat_final].copy()
y = df['PerdidaCartera'].copy()

print("\n=== Shapes y features finales ===")
print("X:", X.shape, "| y rate (1):", y.mean().round(3))
print("Numéricas:", len(num_final), " | Categóricas:", len(cat_final))



>>> ScoreCrediticio: negativos forzados a 0
       ScoreCrediticio  Flag_Score_Negativo
count    146939.000000             146939.0
mean        530.016837                  0.0
std         254.226418                  0.0
min           0.000000                  0.0
25%         521.000000                  0.0
50%         611.000000                  0.0
75%         687.000000                  0.0
max         999.000000                  0.0

ScoreBucket (rehecho):
ScoreBucket
alto        41836
medio       40654
bajo        40652
sin_info    23797

>>> p99 usados para cap:
DiasDesdeUltimoUso                        1102.000
MesesDesdeVinculacion                      116.279
TotalPagosEfectuadosGlobalmentePrevius     219.000
TotalPagosEfectuadosLocalmentePrevius       11.000
NumeroCreditosGPrevius                      76.000
NumeroCreditosGCanalFPrevius                75.000
NumeroCreditosGCanalVPrevius                 5.000
NumeroCreditosGEstadoActivosPrevius          5.000
NumeroCreditosGEs

## 📌 4. Modelo baseline propuesto

**Primera pasada (interpretable):** Logistic Regression con:

- Variables transformadas y estandarizadas/WOE.

- OneHot para categóricas (TipoMunicipioEntregaTC, CanalMunicipioEntregaTC, Genero).

- Reportar ROC-AUC, PR-AUC, KS, curva de beneficio.

**Segunda pasada (mejor performance):** LightGBM con:

- Monotonic constraints en variables como score, ingresos, número de créditos activos.

- Early stopping y validación estratificada.

## 📌 5. Métricas recomendadas para este problema

- ROC-AUC y PR-AUC para medir discriminación.

- KS para ver separación de distribuciones.

- Lift en deciles para ver capacidad de rankear.

- Curva de costo/beneficio: define costo FP y FN para elegir umbral.

- Calibración si necesitas probabilidad real (para pricing/limites).

## Variables con más poder predictivo inmediato

# Un baseline model interpretable

**1899**

In [ ]:
# 1) Parsear fechas
# FechaEvento viene en ISO con 'Z' (UTC). La hacemos tz-aware y luego le quitamos la tz.
clientes['FechaEvento_dt'] = (
    pd.to_datetime(clientes['FechaEvento'], errors='coerce', utc=True)
      .dt.tz_convert(None)  # elimina la tz y deja timestamps "naive"
)

# Fechas que vienen como número de Excel (días desde 1899-12-30)
for c in ['FechaVinculacionCliente','FechaPrimerUso','FechaUltimoUso']:
    clientes[c + '_dt'] = pd.to_datetime(
        clientes[c], errors='coerce', origin='1899-12-30', unit='D'
    )

# 2) Comprobar que todas sean tz-naive
print("FechaEvento_dt tz:", getattr(clientes['FechaEvento_dt'].dt, 'tz', None))
print("FechaVinculacionCliente_dt tz:", getattr(clientes['FechaVinculacionCliente_dt'].dt, 'tz', None))

# 3) Diferencias en días/meses (evitando negativos y nulos)
def safe_months(a, b):
    """(a - b) en meses, con NaN→0 y negativos recortados a 0."""
    delta_days = (a - b).dt.days
    delta_days = delta_days.where(delta_days.notna(), 0)
    delta_days = np.where(delta_days < 0, 0, delta_days)
    return delta_days / 30.0

def safe_days(a, b):
    delta_days = (a - b).dt.days
    delta_days = delta_days.where(delta_days.notna(), 0)
    delta_days = np.where(delta_days < 0, 0, delta_days)
    return delta_days

clientes['MesesDesdeVinculacion'] = safe_months(
    clientes['FechaEvento_dt'], clientes['FechaVinculacionCliente_dt']
)
clientes['MesesDesdePrimerUso'] = safe_months(
    clientes['FechaEvento_dt'], clientes['FechaPrimerUso_dt']
)
clientes['DiasDesdeUltimoUso'] = safe_days(
    clientes['FechaEvento_dt'], clientes['FechaUltimoUso_dt']
)

# 4) Verifica resultados
print(clientes[['FechaEvento_dt','FechaVinculacionCliente_dt','FechaPrimerUso_dt','FechaUltimoUso_dt']].head())
print(clientes[['MesesDesdeVinculacion','MesesDesdePrimerUso','DiasDesdeUltimoUso']].describe())


FechaEvento_dt tz: None
FechaVinculacionCliente_dt tz: None
           FechaEvento_dt FechaVinculacionCliente_dt FechaPrimerUso_dt  \
0 2022-09-19 13:25:31.867                 2019-09-11        2006-12-14   
1 2023-08-23 11:33:46.417                 2023-08-19        2008-07-21   
2 2022-10-01 14:59:48.920                 2019-03-21        2008-04-03   
3 2022-09-22 21:25:09.187                 2022-05-07        2009-07-25   
4 2023-03-19 17:48:52.310                 2021-10-27        2021-10-27   

  FechaUltimoUso_dt  
0        2022-07-23  
1        2023-08-20  
2        2022-09-26  
3        2022-06-04  
4        2022-09-03  
       MesesDesdeVinculacion  MesesDesdePrimerUso  DiasDesdeUltimoUso
count          146939.000000        146939.000000       146939.000000
mean               16.894246            20.877921           90.738762
std                21.224389            28.240460          237.213462
min                 0.000000             0.000000            0.000000
25%          

In [15]:
TARGET = 'PerdidaCartera'

# ---------- 1) USABILIDAD CUPO ----------
# Convertir, cap, flags e imputación
clientes['UsabilidadCupo'] = pd.to_numeric(clientes['UsabilidadCupo'], errors='coerce')
clientes['Flag_Usab_NaN']   = clientes['UsabilidadCupo'].isna().astype(int)
# Cap extremo por reglas de negocio: ratio esperado [0,1], toleramos hasta 2 por seguridad
clientes['Flag_Usab_Outlier'] = ((clientes['UsabilidadCupo'] < 0) | (clientes['UsabilidadCupo'] > 2)).astype(int)
clientes.loc[clientes['UsabilidadCupo'] < 0, 'UsabilidadCupo'] = np.nan
clientes.loc[clientes['UsabilidadCupo'] > 2, 'UsabilidadCupo'] = np.nan
# Imputar con mediana
clientes['UsabilidadCupo'] = clientes['UsabilidadCupo'].fillna(clientes['UsabilidadCupo'].median())

print("\n=== Auditoría UsabilidadCupo ===")
print(clientes[['UsabilidadCupo','Flag_Usab_NaN','Flag_Usab_Outlier']].describe(include='all').to_string())

# ---------- 2) FLAGS DE FECHAS (LEAKAGE) ----------
# Ultimo uso posterior al evento -> flag
clientes['Flag_UltimoUsoPosterior'] = (clientes['FechaUltimoUso_dt'] > clientes['FechaEvento_dt']).fillna(False).astype(int)

print("\n=== Flag_UltimoUsoPosterior (cuenta) ===")
print(clientes['Flag_UltimoUsoPosterior'].value_counts(dropna=False).to_string())

# ---------- 3) NUMÉRICAS CON NULOS: imputar 0 + flag ----------
num_cols_candidates = [
    'DiasMaximosMoraCreditosGenerados',
    'NumeroCreditosGPrevius','NumeroCreditosGCanalFPrevius','NumeroCreditosGCanalVPrevius',
    'NumeroCreditosGEstadoActivosPrevius','NumeroCreditosGEstadoPagadosPrevius',
    'NumeroCreditosLPrevius','NumeroCreditosLEstadoActivosPrevius','NumeroCreditosLEstadoPagadosPrevius',
    'TotalPagosEfectuadosGlobalmentePrevius','TotalPagosEfectuadosLocalmentePrevius',
    'NumeroIntentosFallidos','CupoAprobado','ScoreCrediticio','Edad',
    'MesesDesdeVinculacion','MesesDesdePrimerUso','DiasDesdeUltimoUso'
]

# Crear flags y luego imputar (solo donde tiene NaN)
for c in num_cols_candidates:
    if c in clientes.columns:
        clientes[f'Flag_{c}_NaN'] = clientes[c].isna().astype(int)
        # Para conteos históricos y pagos: NaN ~ sin historial -> 0
        if c.startswith('NumeroCreditos') or c.startswith('TotalPagos'):
            clientes[c] = clientes[c].fillna(0)
        elif c in ['DiasMaximosMoraCreditosGenerados','Edad','NumeroIntentosFallidos']:
            clientes[c] = clientes[c].fillna(0)
        elif c in ['CupoAprobado','ScoreCrediticio']:
            # no imputamos aquí todavía (ver manejo de score/cupo abajo)
            pass

# ---------- 4) SCORE & CUPO ----------
# Score: 0 => sin información
clientes['ScoreSinInfo'] = (clientes['ScoreCrediticio'] == 0).astype(int)
# No imputar el 0 (porque es señal), pero si hay NaN residual, usar mediana de los >0
if clientes['ScoreCrediticio'].isna().any():
    med_pos = clientes.loc[clientes['ScoreCrediticio']>0, 'ScoreCrediticio'].median()
    clientes['ScoreCrediticio'] = clientes['ScoreCrediticio'].fillna(med_pos)

# Cupo: crear versión log
clientes['log_CupoAprobado'] = np.log1p(clientes['CupoAprobado'])
# Si hay NaN residual (0.4%), imputar con mediana log
if clientes['log_CupoAprobado'].isna().any():
    clientes['log_CupoAprobado'] = clientes['log_CupoAprobado'].fillna(clientes['log_CupoAprobado'].median())

print("\n=== Auditoría Score/Cupo ===")
print(clientes[['ScoreCrediticio','ScoreSinInfo','CupoAprobado','log_CupoAprobado']].describe().to_string())

# ---------- 5) CATEGÓRICAS: completar nulos ----------
for c in ['CategoriaPrincipalCredito','UsoAppWeb','Genero','TipoMunicipioEntregaTC','CanalMunicipioEntregaTC']:
    if c in clientes.columns:
        clientes[c] = clientes[c].fillna('Desconocido')

# ---------- 6) CONTROL DE RANGOS / SANITY CHECKS ----------
# Edad plausible
clientes['Flag_Edad_Out'] = (~clientes['Edad'].between(18, 100, inclusive='both')).fillna(False).astype(int)
clientes.loc[clientes['Flag_Edad_Out']==1, 'Edad'] = np.nan
clientes['Edad'] = clientes['Edad'].fillna(clientes['Edad'].median())

# Nunca uses identificadores y fechas crudas como features
drop_from_features = [
    'IdentificadorCliente','FechaEvento','FechaVinculacionCliente','FechaPrimerUso','FechaUltimoUso',
    'FechaEvento_dt','FechaVinculacionCliente_dt','FechaPrimerUso_dt','FechaUltimoUso_dt','FechaPrimerUso_corr',
    'CodigoAlmacenEntregaTC','CodigoAlmacenEntregaTC_str','AlmacenTop20',
    'CodigoMunicipioEntregaTC','MunicipioCat','MunicipioTop20','MesCompra',
    'DiasMora'  # fuga
]
# Asegúrate de crear o tener estas dos si las usas:
if 'Flag_PrimerUsoTemu' not in clientes.columns and 'NumeroCreditosGPrevius' in clientes.columns:
    clientes['Flag_PrimerUsoTemu'] = (clientes['NumeroCreditosGPrevius']==0).astype(int)

# ---------- 7) LISTA FINAL DE FEATURES (baseline) ----------
num_final = [
    'UsabilidadCupo','MesesDesdeVinculacion','MesesDesdePrimerUso','DiasDesdeUltimoUso',
    'NumeroCreditosGPrevius','NumeroCreditosGCanalFPrevius','NumeroCreditosGCanalVPrevius',
    'NumeroCreditosGEstadoActivosPrevius','NumeroCreditosGEstadoPagadosPrevius',
    'NumeroCreditosLPrevius','NumeroCreditosLEstadoActivosPrevius','NumeroCreditosLEstadoPagadosPrevius',
    'TotalPagosEfectuadosGlobalmentePrevius','TotalPagosEfectuadosLocalmentePrevius',
    'NumeroIntentosFallidos','ScoreCrediticio','ScoreSinInfo','CupoAprobado','log_CupoAprobado','Edad',
    'Flag_Usab_NaN','Flag_Usab_Outlier','Flag_PrimerUsoAntesVinc','Flag_UltimoUsoPosterior','Flag_Edad_Out'
] + [f'Flag_{c}_NaN' for c in num_cols_candidates if f'Flag_{c}_NaN' in clientes.columns]

cat_final = [
    'Genero','TipoMunicipioEntregaTC','CanalMunicipioEntregaTC','UsoAppWeb','CategoriaPrincipalCredito',
    'Flag_PrimerUsoTemu'  # binaria, si prefieres tratarla como numérica puedes dejarla en num_final
]

# Quitar columnas ausentes o a excluir
num_final = [c for c in num_final if c in clientes.columns and c not in drop_from_features]
cat_final = [c for c in cat_final if c in clientes.columns and c not in drop_from_features]

X = clientes[num_final + cat_final].copy()
y = clientes[TARGET].copy()

print("\n=== Resumen columnas finales ===")
print("Numéricas (", len(num_final), "):", num_final[:12], "...")
print("Categóricas (", len(cat_final), "):", cat_final)

# ---------- 8) AUDITORÍA POST-LIMPIEZA ----------
# Correlación numérica con el target
print("\n=== Correlación numéricas (post-limpieza) con PerdidaCartera ===")
print(X[num_final].corrwith(y).sort_values(ascending=False).to_string())

# Tablas cruzadas para categóricas
for c in cat_final:
    if X[c].nunique() <= 40:
        print(f"\n=== {c} vs {TARGET} ===")
        print(pd.crosstab(X[c], y, normalize='index').round(3).to_string())

# Balance de clases (importante para métricas a usar)
print("\n=== Balance de clases ===")
print(y.value_counts(normalize=True).rename({0:'% no pérdida',1:'% pérdida'}).round(3).to_string())



=== Auditoría UsabilidadCupo ===
       UsabilidadCupo  Flag_Usab_NaN  Flag_Usab_Outlier
count   146939.000000  146939.000000      146939.000000
mean         0.351514       0.086778           0.058398
std          0.256366       0.281510           0.234496
min          0.002082       0.000000           0.000000
25%          0.166692       0.000000           0.000000
50%          0.278900       0.000000           0.000000
75%          0.498059       0.000000           0.000000
max          0.999995       1.000000           1.000000

=== Flag_UltimoUsoPosterior (cuenta) ===
Flag_UltimoUsoPosterior
0    146939

=== Auditoría Score/Cupo ===
       ScoreCrediticio   ScoreSinInfo  CupoAprobado  log_CupoAprobado
count    146939.000000  146939.000000  1.462790e+05     146939.000000
mean        530.002688       0.158569  1.395009e+10         22.934600
std         254.256067       0.365275  1.151370e+10          0.983373
min          -9.000000       0.000000  1.000000e+09         20.723266
25% 